### PR TITLE
Non-local components and prep for publishing

### PR DESCRIPTION
**Pull request by nickjalbert**
This tackles the [TODO list](https://github.com/agentos-project/agentos/pull/188#issuecomment-965995156) from the previous PR as well as adds support for:

* Components from a Registry (currently the Registry is just mocked out by a static dict)
* Components from GitHub
* Support for publishing a run in the Tracker component

## Demo Registry

Run the following to pull the SB3 Agent from the registry (the registry is a mocked local dictionary, the agent is coming from github):

```python
# Run your Python interpreter in a clean dir
>>> import agentos
>>> component = agentos.get_component('sb3_ppo_agent')'
>>> component.run('learn')
>>> component.run('evaluate')
>>> component = agentos.get_component('sb3_tracker')
>>> component.run('publish')  # prints the run data we'd publish
```

I get the following on my machine:

```bash
>>> import agentos
>>> component = agentos.get_component('sb3_ppo_agent')
Cloning into '/home/nickj/.agentos_cache/nickjalbert/agentos/nj_registry_2next'...
remote: Enumerating objects: 3453, done.
remote: Counting objects: 100% (1838/1838), done.
remote: Compressing objects: 100% (1032/1032), done.
remote: Total 3453 (delta 1150), reused 1389 (delta 761), pack-reused 1615
Receiving objects: 100% (3453/3453), 10.34 MiB | 8.35 MiB/s, done.
Resolving deltas: 100% (2057/2057), done.
>>> component.run('learn')
Adding environment to sb3_ppo_agent
Calling sb3_cartpole.__init__(**{})
Adding tracker to sb3_ppo_agent
Calling sb3_tracker.__init__(**{})
Calling sb3_ppo_agent.__init__(**{})
SB3Tracker: No saved SB3 PPO 'ppo' found.
Calling sb3_ppo_agent.learn(**{})

Training results over 96 episodes:
        Overall agent was trained on 2048 transitions over 96 episodes
        Max reward over 96 episodes: 100.0
        Mean reward over 96 episodes: 21.333333333333332
        Median reward over 96 episodes: 17.0
        Min reward over 96 episodes: 6.0

>>> component.run('evaluate')
Adding environment to sb3_ppo_agent
Calling sb3_cartpole.__init__(**{})
Adding tracker to sb3_ppo_agent
Calling sb3_tracker.__init__(**{})
Calling sb3_ppo_agent.__init__(**{})
SB3Tracker: Restored SB3 PPO model 'ppo'.
Calling sb3_ppo_agent.evaluate(**{})
CartPole: Reached 500 steps per episode limit!  Truncating episode.

Benchmark results over 10 episodes:
        Benchmarked agent was trained on 2048 transitions over 96 episodes
        Max reward over 10 episodes: 501.0
        Mean reward over 10 episodes: 147.2
        Median reward over 10 episodes: 74.5
        Min reward over 10 episodes: 56.0

>>> component = agentos.get_component('sb3_tracker')
>>> component.run('publish')
Calling sb3_tracker.__init__(**{})
Calling sb3_tracker.publish(**{})
{'artifact_paths': [PosixPath('/home/nickj/agentos/demo/mlruns/0/f4e5124c7a4c4fe7a11720093b9d2895/artifacts/ppo.zip')],
 'component_spec': {'components': {'sb3_cartpole==nj_registry_2next': {'class_name': 'CartPole',
                                                                       'dependencies': {},
                                                                       'file_path': 'example_agents/sb3_agent/environment.py',
                                                                       'repo': 'dev_repo'},
                                   'sb3_ppo_agent==nj_registry_2next': {'class_name': 'SB3PPOAgent',
                                                                        'dependencies': {'environment': 'sb3_cartpole==nj_registry_2next',
                                                                                         'tracker': 'sb3_tracker==nj_registry_2next'},
                                                                        'file_path': 'example_agents/sb3_agent/agent.py',
                                                                        'repo': 'dev_repo'},
                                   'sb3_tracker==nj_registry_2next': {'class_name': 'SB3Tracker',
                                                                      'dependencies': {},
                                                                      'file_path': 'example_agents/sb3_agent/tracker.py',
                                                                      'repo': 'dev_repo'}},
                    'repos': {'dev_repo': {'type': 'github',
                                           'url': 'https://github.com/nickjalbert/agentos'}}},
 'entry_point': 'evaluate',
 'is_published': 'True',
 'metrics': {'episode_count': 10.0,
             'max_reward': 501.0,
             'mean_reward': 147.2,
             'median_reward': 74.5,
             'min_reward': 56.0,
             'step_count': 1472.0,
             'training_episode_count': 96.0,
             'training_step_count': 2048.0},
 'parameters': {}}
```

## Demo GitHub components

Run the following from your command-line:

```bash
cd example_agents/gh_sb3_agent
# Note the agentos.yaml file in this directory
agentos run agent --entry-point learn
agentos run agent --entry-point evaluate
agentos run  tracker --entry-point  publish
```

### CODE REVIEW

**Comment by nickjalbert** [comment]
Re our discussion this morning:  [dulwich](https://www.dulwich.io/docs/index.html) is a Python implementation of git that's pip installable and importable.  Tested it out on the REPL and seems like we could use this instead of relying on git to be installed locally.  I think I'll hold off on this for now, though, and work on the end-to-end stuff.

File: agentos/component.py [hunk 1]
```diff
@@ -223,3 +141,111 @@ def _parse_components(cls, components_spec: Dict, repos: Dict) -> Dict:
                 component.add_dependency(dependency, alias=alias)
 
         return components
+
+    def run(self, fn_name: str, params: ParameterSet = None):
+        params = params if params else ParameterSet()
+        with self._track_run(fn_name, params):
+            return self._untracked_run(fn_name, params)
+
+    def _untracked_run(
```
Comment (andyk) [inline agentos/component.py:150]: As we discussed in-person, can we get rid of `_untracked_run()` and instead have `run()` take a `tracked=False` flag?

File: agentos/component.py [hunk 2]
```diff
@@ -223,3 +141,111 @@ def _parse_components(cls, components_spec: Dict, repos: Dict) -> Dict:
...
+            )
+            setattr(instance, dep_alias, dep_instance)
+        setattr(instance, self._dunder_name, self)
+        self._managed_cls.__init__ = save_init
+        self._untracked_run("__init__", params=params, instance=instance)
+        instantiated[self.name] = instance
+        return instance
+
+    def _log_params(self, params: ParameterSet) -> None:
+        log_data_as_yaml_artifact("parameters.yaml", params.to_dict())
```
Comment (andyk) [inline agentos/component.py:191]: Can we call it `component_parameter_set.yaml` or just `parameter_set.yaml` instead of `parameter.yaml`?

**Review by andyk — COMMENTED** [review]
Requesting mostly small changes.

One final optional thing that would be helpful if you feel up to it is to add a doc string at the class level of each of the main core components (Repo, ParameterSet) if they don't have one currently.

**Comment by andyk** [comment]
> Re our discussion this morning:  [dulwich](https://www.dulwich.io/docs/index.html) is a Python implementation of git that's pip installable and importable.  Tested it out on the REPL and seems like we could use this instead of relying on git to be installed locally.  I think I'll hold off on this for now, though, and work on the end-to-end stuff.

Made #198

**Comment by nickjalbert** [comment]
I was able to fix all nits (including removing `runtime.py` and using that Python implementation of git).   Will merge once tests are green!

**Pull request merged**